code/modules/paperwork/paper_bundle.dm

```diff
@@ -98,145 +197,354 @@
-		dat+= "<DIV STYLE='float:left; text-align:center; width:33.33333%'><A href='?src=\ref[src];remove=1'>Remove [(istype(W, /obj/item/paper)) ? "paper" : "photo"]</A></DIV>"
-		dat+= "<DIV STYLE='float;left; text-align:right; with:33.33333%'><A href='?src=\ref[src];next_page=1'>Back</A></DIV><BR><HR>"
-	// middle pages
+	var/obj/item/W = pages[cur_page]
+
+	//Header
+	dat = "<TABLE STYLE='white-space:nowrap; overflow:clip; width:100%; height:2em; table-layout:fixed;'><TR>"
+	dat += "<TD style='text-align:center;'>"
+	if(cur_page > 1)
+		dat += "<A href='?src=\ref[src];first_page=1'>First</A>"
+	else
+		dat += "First"
+	dat += "</TD>"
+
+	dat += "<TD style='text-align:center;'>"
+	if(cur_page > 1)
+		dat += "<A href='?src=\ref[src];prev_page=1'>Previous</A>"
+	else
+		dat += "Previous"
+	dat += "</TD>"
+
+	dat += "<TD style='text-align:center;'><A href='?src=\ref[src];jump_to=1;'><B>[cur_page]/[length(pages)]</B></A> <A href='?src=\ref[src];remove=1'>Remove</A></TD>"
+
+	dat += "<TD style='text-align:center;'>"
+	if(cur_page < pages.len)
+		dat += "<A href='?src=\ref[src];next_page=1'>Next</A>"
+	else
+		dat += "Next"
+	dat += "</TD>"
+
+	dat += "<TD style='text-align:center;'>"
+	if(cur_page < pages.len)
+		dat += "<A href='?src=\ref[src];last_page=1'>Last</A>"
 	else
-		dat+= "<DIV STYLE='float:left; text-align:left; width:33.33333%'><A href='?src=\ref[src];prev_page=1'>Previous Page</A></DIV>"
-		dat+= "<DIV STYLE='float:left; text-align:center; width:33.33333%'><A href='?src=\ref[src];remove=1'>Remove [(istype(W, /obj/item/paper)) ? "paper" : "photo"]</A></DIV>"
-		dat+= "<DIV STYLE='float:left; text-align:right; width:33.33333%'><A href='?src=\ref[src];next_page=1'>Next Page</A></DIV><BR><HR>"
+		dat += "Last"
+	dat += "</TD>"
+	dat += "</TR></TABLE><HR>"
 
-	if(istype(pages[page], /obj/item/paper))
+	//Contents
+	if(istype(W, /obj/item/paper))
 		var/obj/item/paper/P = W
-		dat+= "<HTML><HEAD><TITLE>[P.name]</TITLE></HEAD><BODY>[P.show_info(user)][P.stamps]</BODY></HTML>"
+		dat += "<HTML><HEAD><TITLE>[P.name]</TITLE></HEAD><BODY>[P.info][P.stamp_text]</BODY></HTML>"
 		show_browser(user, dat, "window=[name]")
-	else if(istype(pages[page], /obj/item/photo))
+		onclose(user, name)
+
+	else if(istype(W, /obj/item/photo))
 		var/obj/item/photo/P = W
-		dat += "<html><head><title>[P.name]</title></head><body style='overflow:hidden'>"
-		dat += "<div> <img src='tmp_photo.png' width = '180'[P.scribble ? "<div> Written on the back:<br><i>[P.scribble]</i>" : null ]</body></html>"
+		dat += {"
+			<html><head><title>[P.name]</title></head><body style='overflow:hidden'>
+			<div> <img src='tmp_photo.png' width = '180'[P.scribble ? "<div> Written on the back:<br><i>[P.scribble]</i>" : null ]</body></html>
+		"}
 		send_rsc(user, P.img, "tmp_photo.png")
-		show_browser(user, JOINTEXT(dat), "window=[name]")
+		show_browser(user, dat, "window=[name]")
+		onclose(user, name)
+	user.set_machine(src)
```

Comment:
Don't do this manually. Also: what do we need onclose for?

Reply:
Look at how set_machine works. Then look at what the verb called by byond when onclose is set does.